### ISSUE TITLE
Comparison in loop generates slightly worse code than GCC

### ISSUE THREAD

**Issue by firewave** [llvm/llvm-project#60761]
Apologies in advance when I am on the wrong track here. I still (and probably always will) have issues with understanding byte code properly.

```cpp
int multiCompare1(const char *tok, const char *haystack)
{
    const char *needle = tok;
    const char *needlePointer = needle;
    for (;;) {
        if (*haystack == ' ' || *haystack == '\0') {
            if (needlePointer == needle)
                return 0;
            break;
        }
        else {
            do {
                ++haystack;

                if (*haystack == ' ' || *haystack == '\0') { // <---
                    return -1;
                }
                if (*haystack == '|') {
                    break;
                }
            } while (true);

            ++haystack;
        }
    }

    return -1;
}
```

If I replace the marked `if` with a `switch` it will generate one less branch. If I pull out the code and try to get a more reduced case it will generate the same code for both cases. If you remove the `if` after that check it will also generate the "expected" code.

Using the `switch` will match the code GCC generates for the `if` case. But with `switch` GCC will generate different code.

https://godbolt.org/z/9bbdfWejh

If I reduce the code too much GCC is also able to recognize that the loop won't do anything: https://godbolt.org/z/YMEz46hdE. That's probably a different ticket though.

**Comment by firewave** [comment]
Looking at the IR it appears the `if` version is better.

Something curious about the `switch` version is that it does a `sext` to `i32` which looks unnecessary.

**Comment by firewave** [comment]
https://godbolt.org/z/GKobTq6n5

with `if`: 
```
6:                                                ; preds = %3, %6
  %7 = phi ptr [ %8, %6 ], [ %4, %3 ]
  %8 = getelementptr inbounds i8, ptr %7, i64 1
  %9 = load i8, ptr %8, align 1, !tbaa !6
  switch i8 %9, label %6 [
    i8 32, label %12
    i8 0, label %12
    i8 124, label %10
  ]
```

with `switch:`
```
7:                                                ; preds = %4, %12
  %8 = phi ptr [ %9, %12 ], [ %5, %4 ]
  %9 = getelementptr inbounds i8, ptr %8, i64 1
  %10 = load i8, ptr %9, align 1, !tbaa !6
  %11 = sext i8 %10 to i32
  switch i32 %11, label %12 [
    i32 32, label %16
    i32 0, label %16
  ]
```

**Comment by nikic** [comment]
Yeah, that sext prevents the fallthrough condition from being merged into the switch.

**Comment by caojoshua** [comment]
Agree with nikic. SimplifyCFG would fold the branch into the switch if the branch and switch are comparing against the same value against a condition [[source](https://github.com/llvm/llvm-project/blob/70d78e035801a35c95c55384621724b0ef8ecb35/llvm/lib/Transforms/Utils/SimplifyCFG.cpp#L1331)]. But the values are not the same, because the switch has a sext value.

It looks like clang is generating the sext value. I'm curious why the condition needs to be sext'ed before being used by the switch.

I don't think this is a case LLVM should handle. LLVM is already doing the right thing by checking that folding branch/switch is legal if they have the same condition. And I'm not sure it will always be better to check that conditions are equal only after a sext.
